analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs

```diff
@@ -18,16 +18,55 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
 namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks.CSharp;
 
 public class PublicMethodArgumentsShouldBeCheckedForNull : SymbolicRuleCheck
 {
     private const string DiagnosticId = "S3900";
-    private const string MessageFormat = "Refactor this method to add validation of parameter '{0}' before using it.";
+    private const string MessageFormat = "{0}";
 
     internal static readonly DiagnosticDescriptor S3900 = DescriptorFactory.Create(DiagnosticId, MessageFormat);
 
     protected override DiagnosticDescriptor Rule => S3900;
 
-    public override bool ShouldExecute() => false;
+    public override bool ShouldExecute() => true;
+
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        if (context.Operation.Instance.Kind == OperationKindEx.ParameterReference
+            && context.Operation.Instance.ToParameterReference() is { WrappedOperation: var reference, WrappedOperation.Syntax: var syntax, Parameter: var parameterSymbol }
+            && !parameterSymbol.Type.IsValueType
+            && IsParameterDereferenced(context.Operation)
+            && !IgnoreBecauseOfParameterAttribute(parameterSymbol)
+            && NullableStateIsNotKnownForParameter(parameterSymbol)
+            && SemanticModel.GetDeclaredSymbol(Node) is { } methodSymbol)
```

Comment:
In which cases is this condition false? You should move the `GetDeclaredSymbol` to the if body if the answer is none.

Reply:
It can be null for local functions and lambda expressions. The proper implementation for the `ShouldExecute` method is in a [seperate PR](https://github.com/SonarSource/sonar-dotnet/pull/6942), but - for now - I've added a basic check to filter out lambdas/local functions. Then I moved the `IsConstructor` check to the body.